src/mcpwm/timer.rs

```diff
@@ -0,0 +1,302 @@
+
+        // TODO: What should these be set to?
+        flags.set_update_period_on_empty(1);
+        flags.set_update_period_on_sync(0);
+
+        let cfg = mcpwm_timer_config_t {
+            group_id: G::ID,
+            clk_src: soc_periph_mcpwm_timer_clk_src_t_MCPWM_TIMER_CLK_SRC_DEFAULT,
+            resolution_hz: config.resolution.0,
+            count_mode: config.count_mode.into(),
+            period_ticks: config.period_ticks.into(),
+            flags,
+        };
+        let mut handle: mcpwm_timer_handle_t = ptr::null_mut();
+        unsafe {
+            esp!(mcpwm_new_timer(&cfg, &mut handle)).unwrap();
+        }
+        // TODO: note that this has to be called before mcpwm_timer_enable
+        // mcpwm_timer_register_event_callbacks()
+        unsafe {
+            esp!(mcpwm_timer_enable(handle)).unwrap();
+            esp!(mcpwm_timer_start_stop(
+                handle,
+                mcpwm_timer_start_stop_cmd_t_MCPWM_TIMER_START_NO_STOP
+            ))
+            .unwrap();
+        }
+
+        let period_peak = if config.count_mode == CountMode::UpDown {
+            (cfg.period_ticks / 2).try_into().unwrap()
+        } else {
+            cfg.period_ticks.try_into().unwrap()
+        };
+
+        Self {
+            _group: G::default(),
+            handle,
+            _timer: timer,
+            period_ticks: cfg.period_ticks,
+            period_peak,
+        }
+    }
+
+    // TODO: make sure this description is accurate
+    /// Get number of ticks per period
+    ///
+    /// Use this when working with the frequency or the period
+    ///
+    /// NOTE: This will be the same as `Self::get_period_peak` for all `CounterMode` except for
+    /// `CounterMode::UpDown` where the period will be twice as large as the peak value since
+    /// the timer will count from zero to peak and then down to zero again
+    pub fn get_period_ticks(&self) -> u32 {
+        self.period_ticks
+    }
+
+    // TODO: make sure this description is accurate
+    /// This is the maximum value that the comparator will see
+    ///
+    /// Use this working with the duty
+    ///
+    /// NOTE: This will not be the same as `Self::get_period_ticks` when using `CounterMode::UpDown`
+    /// See `Self::get_period_ticks` for more info
+    pub fn get_period_peak(&self) -> u16 {
+        self.period_peak
+    }
+
+    pub fn timer(&self) -> mcpwm_timer_handle_t {
+        self.handle
+    }
+
+    // TODO: It seems that we can't have both at the same time:
+    // a method for releasing its hardware resources
+    // and implementing Drop.
+    /*
+    pub fn release(self) -> TIMER<N, G> {
+        let Self {
+            _group,
+            _timer,
+            handle
+        } = self;
+        unsafe {
+            esp!(mcpwm_del_timer(handle)).unwrap();
+        }
+        _timer
+    }*/
+
+    pub fn into_connection(self) -> TimerConnection<N, G, NoOperator, NoOperator, NoOperator> {
```

Comment:
I have some concerns around this method being consuming, but let's discuss this later, after we have covered some of the other changes I've suggested.